### ISSUE TITLE
[Air|Train] `TorchTrainer` cannot train ImageNet model because label becomes zero

### ISSUE THREAD

**Issue by Tonyhao96** [ray-project/ray#30168]
### What happened + What you expected to happen

I follow the Ray Train example in [torch_fashion_mnist_example](https://github.com/ray-project/ray/blob/master/python/ray/train/examples/pytorch/torch_fashion_mnist_example.py). It works well on MNIST and CIFAR-10 datasets with multiple GPUs. (Model ResNet-18) 

But when I use the ImageNet dataset also using ResNet-18, the accuracy always keeps zero. I don't change other training code part. After debugging, I find the label becomes zero using ray:

### Ray (The last image data \& Label)

```bash
(RayTrainWorker pid=272320) 
(RayTrainWorker pid=272320)          [[-0.8807, -0.8110, -0.6367,  ...,  0.5834,  0.8099,  0.8274],
(RayTrainWorker pid=272320)           [-0.8981, -0.8284, -0.6890,  ...,  0.3393,  0.6705,  0.8971],
(RayTrainWorker pid=272320)           [-0.8284, -0.7761, -0.7238,  ...,  0.2348,  0.2173,  0.7054],
(RayTrainWorker pid=272320)           ...,
(RayTrainWorker pid=272320)           [-0.7413, -0.4798, -0.7413,  ...,  0.2348,  0.2522,  0.1825],
(RayTrainWorker pid=272320)           [-0.7936, -0.5147, -0.7238,  ...,  0.3742,  0.3916,  0.3568],
(RayTrainWorker pid=272320)           [-0.8110, -0.4798, -0.7413,  ...,  0.3742,  0.4265,  0.4439]]]])
(RayTrainWorker pid=272320) tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
(RayTrainWorker pid=272320)         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
(RayTrainWorker pid=272320)         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
(RayTrainWorker pid=272320)         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
(RayTrainWorker pid=272320)         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
(RayTrainWorker pid=272320)         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
(RayTrainWorker pid=272320)         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
(RayTrainWorker pid=272320)         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
(RayTrainWorker pid=272320)         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
(RayTrainWorker pid=272320)         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
(RayTrainWorker pid=272320)         0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
```

### Pytorch Official Example (The last image data \& Label)
```bash
         [[ 1.1062,  1.0714,  1.0714,  ...,  1.3502,  1.4025,  1.5071],
          [ 1.0365,  0.9319,  0.9145,  ...,  1.3328,  1.3677,  1.5071],
          [ 1.1237,  0.9668,  0.8971,  ...,  1.2980,  1.3851,  1.5071],
          ...,
          [ 0.7751,  1.3154,  1.3677,  ..., -1.1421, -0.1661,  0.6705],
          [ 0.5834,  0.8099,  1.1759,  ..., -0.8110,  0.9145,  1.1062],
          [ 2.0474,  1.8731,  1.8905,  ..., -0.8110,  0.2871, -0.0267]]]])
tensor([445, 177, 440, 460,  60, 748, 792, 106, 518, 522, 768, 903, 932, 556,
         58, 687, 997, 703, 349,  40, 355, 848, 730, 752, 837, 827,  31, 142,
        759, 513, 239, 485, 648, 500, 608, 367, 798,  14, 345, 338,  66, 965,
        540, 624, 662, 371, 892, 607, 549, 931, 881, 239, 176, 366, 341, 142,
        621, 827, 564,  46, 880, 983, 889, 375, 649,  18, 909, 149, 748, 339,
        674, 623, 713, 871, 468, 441, 474, 310, 496, 401, 728, 633, 757, 947,
        949, 593, 883,  89, 927, 777, 754, 315, 226, 250, 295, 227, 552, 103,
        147, 195, 392, 283, 910, 957, 711, 452, 671, 688,   7, 834, 186, 773,
         18, 144, 451, 996,  67, 430, 328, 313, 439, 605, 570, 525, 221, 639,
        202, 269, 903, 457, 372, 513, 819, 368, 460, 262, 481, 573, 874, 360,
        497, 630,  79, 686, 575, 753, 937, 674, 102, 434, 726, 908,  72,  87,
        783, 570, 593, 190, 691, 300, 108, 139, 897, 561, 800, 392, 858, 879,
        271, 494, 664, 164, 300, 784, 546, 227, 393, 438, 692, 122,  84, 466,
        312, 165, 965, 171, 798, 661, 101, 680, 110, 352, 761, 127, 387, 253,
         15, 770, 975,  73, 925, 481, 706, 275, 168, 127, 856, 151,  77, 987,
        699, 151, 247, 844, 864, 141, 616, 680, 245, 515, 372, 606, 545, 948,
        725, 124, 769, 851, 385, 831, 566, 317, 876, 843, 468, 153, 206, 120,
        305, 987, 357, 377, 256, 598, 752, 136,  91, 623, 588, 384, 933, 983,
        624, 579, 951, 307])
```

### Versions / Dependencies

Ray 2.1
Pytorch 1.13
Python 3.9
Ubuntu 20.04

### Reproduction script

I just use the same dataloader code with the official code

```Python
    train_set, val_set = get_datasets(dataset)
    train_loader = DataLoader(train_set, batch_size=worker_batch_size, num_workers=8, pin_memory=True)
    val_loader = DataLoader(val_set, batch_size=worker_batch_size, num_workers=8, pin_memory=True)

    images, target = train_loader.__iter__().__next__()
    print("target", target.shape)
    print(images)
    print(target)
```

### Issue Severity

High: It blocks me from completing my task.

**Comment by matthewdeng** [comment]
Hey @Tonyhao96 do you have a more extensive repro I can use to validate this on my end?

One idea I was wondering if the input Dataset comes in sorted by label, but that wouldn't explain why you're seeing different behavior with/without Ray or why accuracy is zero.

**Comment by Tonyhao96** [comment]
Sorry, I found that I missed the shuffle.
